### PR TITLE
[DSV4.1] DeepSeek-V4.1 support prefill CP

### PR DESCRIPTION
**Pull request by zhangxiaolei123456**
## Motivation

Baseline: https://github.com/sgl-project/sglang/pull/38798

DeepSeek-V4.1 DSpark PD disaggregation currently requires Prefill and Decode to use the same attention TP layout, which prevents Prefill from using context parallelism for long prompts.

This PR enables the following topology:

- Prefill: TP8 world, attention TP1 x CP8
- Decode: attention TP8 x CP1
- Mooncake PD transfer
- DSpark static verification

## Modifications

- Allow canonical interleave Prefill CP for DeepSeek-V4.1 DSpark PD.
- Validate equal total attention parallel width instead of requiring identical `attn_tp_size`.
- Connect each Decode rank to all Prefill CP ranks.
- Partition globally materialized compressed KV and Indexer pages across CP senders.
- Transfer CP-replicated state only from CP rank 0, including target SWA, draft SWA, and the C2 request ring.
- Preserve the existing odd-boundary whole-request C2 ring handoff.
- Fix CP-v2 decoder bounded replay:
  - use tail-specific CP metadata for hidden-state gathering;
  - preserve global token indices for DSpark hidden injection;
  - use tail-specific input IDs and logits metadata.
- Allow text requests from DeepSeek-V4.1 multimodal checkpoints to use Prefill CP while continuing to reject multimodal requests under CP.
- Align Engram image-token masking with CP-local input IDs.
- Rebuild global routing IDs when entering the bounded-replay tail.
- Fall back to the existing Torch indexer on Hopper/SM90, since DeepGEMM FP8-FP4 MQA logits only supports SM100/SM120.

## Supported Configuration

- `--enable-prefill-cp`
- `--cp-strategy interleave`
- Prefill CP only; Decode must use CP1
- DP1, DCP1, PP1
- Mooncake transfer backend
- `SGLANG_RAGGED_VERIFY_MODE=static`
- DSpark enabled on both Prefill and Decode
- Matching DSpark block size and KV layout on both sides
- Text-only requests under CP

The following remain unsupported:

- Decode CP/DCP
- Multimodal requests with CP
- Encoder SWA bounded replay with CP or PD
- NIXL for this topology

## Accuracy Tests
```
#prefill
SGLANG_RAGGED_VERIFY_MODE=static GLOO_SOCKET_IFNAME=eth0 NCCL_MIN_NCHANNELS=24 NCCL_IB_QPS_PER_CONNECTION=8 sglang serve --trust-remote-code --model-path /data02/models/DeepSeek-V4.1-Flash --tp 8 --ep-size 8 --attention-backend dsv4 --moe-runner-backend flashinfer_mxfp4 --speculative-algorithm DSPARK --enable-decoder-swa-bounded-replay --cuda-graph-backend-prefill disabled --disaggregation-mode prefill --disaggregation-transfer-backend mooncake --disaggregation-bootstrap-port 8998 --disaggregation-ib-device "mlx5_1,mlx5_2,mlx5_3,mlx5_4" --reasoning-parser auto --tool-call-parser auto --host 0.0.0.0 --port 30000 --flashinfer-mxfp4-moe-precision fp8 --mem-fraction-static 0.85 --speculative-dspark-block-size 5 --attn-cp-size 8 --enable-prefill-cp --cp-strategy interleave --max-running-requests 8

#decode
SGLANG_RAGGED_VERIFY_MODE=static GLOO_SOCKET_IFNAME=eth0 NCCL_MIN_NCHANNELS=24 NCCL_IB_QPS_PER_CONNECTION=8 sglang serve --trust-remote-code --model-path /data02/models/DeepSeek-V4.1-Flash --tp 8 --ep-size 8 --attention-backend dsv4 --moe-runner-backend flashinfer_mxfp4 --speculative-algorithm DSPARK --speculative-dspark-block-size 5 --disaggregation-mode decode --disaggregation-transfer-backend mooncake --disaggregation-bootstrap-port 8998 --disaggregation-ib-device "mlx5_1,mlx5_2,mlx5_3,mlx5_4" --reasoning-parser auto --tool-call-parser auto --host 0.0.0.0 --port 30000 --flashinfer-mxfp4-moe-precision fp8 --mem-fraction-static 0.8 --disable-flashinfer-autotune --cuda-graph-max-bs-decode 32 --max-running-requests 32
```
GSM8K
```
sgl-eval run gsm8k --base-url http://localhost:8090/v1 --num-threads 16 --temperature 0.7 --top-p 0.9 --seed 0 --max-tokens 2048 --no-thinking
Run directory: /root/.sgl_eval/sgl_eval_gsm8k_20260912-043851
gsm8k: 100%|█| 1319/1319 [14:02<00:00,  1.57it/
== gsm8k ==
1319 examples (single-shot)  |  842.3s  |  202 tok/s  |  170K tokens
* score           =  96.44%
  stop_rate       =  100.00%
  truncated_rate  =  0.00%
  error_rate      =  0.00%
```

## Speed Tests and Profiling

PR | Hardward | input | output | chunked-prefill-size | QPS | max_concurrency | num promot | TTFT(ms) | throughput(tokens/s/gpu)
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
CP+TP | H20 | 8000 | 1 | 8192 | 2 | 64 | 128 | 4152.92 | 1913.08125
TP | H20 | 8000 | 1 | 8192 | 2 | 64 | 128 | 37106.32 | 1197.21375

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #35177933802](https://github.com/sgl-project/sglang/actions/runs/35177933802)
Latest PR Test (Extra): :x: [Run #35177933619](https://github.com/sgl-project/sglang/actions/runs/35177933619)
Latest PR Test (AMD ROCm 10): :x: [Run #35177933765](https://github.com/sgl-project/sglang/actions/runs/35177933765)

### CODE REVIEW

**Comment by zhangxiaolei123456** [comment]
@Jiminator The dsv4.1 base branch has been restored and #38798 is open again. Could you please reopen this PR? The head branch is still available.
